src/pretix/plugins/paypal/payment.py

```diff
@@ -56,24 +68,77 @@ def settings_form_fields(self):
+                 forms.ChoiceField(
+                     label=_('Endpoint'),
+                     initial='live',
+                     choices=(
+                         ('live', 'Live'),
+                         ('sandbox', 'Sandbox'),
+                     ),
+                 )),
             ] + list(super().settings_form_fields.items())
         )
+
         d.move_to_end('_enabled', False)
         return d
 
+    def get_connect_url(self, request):
+        request.session['payment_paypal_oauth_event'] = request.event.pk
+
+        self.init_api()
+        return Tokeninfo.authorize_url({'scope': 'openid profile email'})
+
     def settings_content_render(self, request):
-        return "<div class='alert alert-info'>%s<br /><code>%s</code></div>" % (
-            _('Please configure a PayPal Webhook to the following endpoint in order to automatically cancel orders '
-              'when payments are refunded externally.'),
-            build_global_uri('plugins:paypal:webhook')
-        )
+        if self.settings.connect_client_id and not self.settings.secret_key:
+            # Use PayPal connect
+            if not self.settings.connect_user_id:
+                return (
+                    "<p>{}</p>"
+                    "<a href='{}' class='btn btn-primary btn-lg'>{}</a>"
+                ).format(
+                    _('To accept payments via PayPal, you will need an account at PayPal. By clicking on the '
+                      'following button, you can either create a new PayPal account connect pretix to an existing '
+                      'one.'),
+                    self.get_connect_url(request),
+                    _('Connect with <i class="fa fa-paypal"></i> PayPal')
+                )
+            else:
+                return (
+                    "<button formaction='{}' class='btn btn-danger'>{}</button>"
+                ).format(
+                    reverse('plugins:paypal:oauth.disconnect', kwargs={
+                        'organizer': self.event.organizer.slug,
+                        'event': self.event.slug,
+                    }),
+                    _('Disconnect from PayPal')
+                )
+        else:
+            return "<div class='alert alert-info'>%s<br /><code>%s</code></div>" % (
+                _('Please configure a PayPal Webhook to the following endpoint in order to automatically cancel orders '
+                  'when payments are refunded externally.'),
+                build_global_uri('plugins:paypal:webhook')
+            )
 
     def init_api(self):
-        paypalrestsdk.set_config(
-            mode="sandbox" if "sandbox" in self.settings.get('endpoint') else 'live',
-            client_id=self.settings.get('client_id'),
-            client_secret=self.settings.get('secret'))
+        if self.settings.connect_client_id:
+            paypalrestsdk.set_config(
+                mode="sandbox" if "sandbox" in self.settings.get('endpoint') else 'live',
```

Comment:
Where is this ``endpoint`` setting set? As far as I can see it, it's neither part of the global settings nor of the event-level settings when OpenID Connect is in use?

Reply:
I guess this only worked for me with the sandbox env, because I had PayPal manually set up before making my changes. Welp...

But I am not 100% sure how we should handle the whole thing. So the following paragraphs might be a little bit confusing.

| Global Setting | Customer can choose to log into | If Customer logged into | Then Customer can choose to do transactions in |
|----------------|---------------------------------|-------------------------|------------------------------------------------|
| Sandbox        | Sandbox only                    | Sandbox                 | Sandbox only                                   |
| Production     | Sandbox or Production           | Sandbox                 | Sandbox only                                   |
| Production     | Sandbox or Production           | Production              | Production only                                |

Given this is more complex than Stripe, I am not 100% sure where we should offer the env-selection.

One issue seems to be, that the accounts in production/sandbox are not the same. So even if a customer logged into production, and then switched to sandbox, things would break.

So here is an idea: 
- Have a env-switch in  the global settings.
- If global == production, show env-switch to customer before pairing with PayPal.
- If global == sandbox, customer-env is forcibly set to sandbox before pairing with PayPal.
- After Pairing, event-level env cannot be changed. In order to switch env, customer has to un- and relink PayPal.

What are your thoughts on that? Too over-engineered?